paddle/fluid/framework/ir/mkldnn/fc_elementwise_add_mkldnn_fuse_pass.cc

```diff
@@ -0,0 +1,319 @@
+    std::stringstream msg_ss;
+    msg_ss << "---    Fused " << found_fc_as_y_count
+           << " fc (as y) + elementwise_add patterns";
+    paddle::string::PrettyLogDetail(msg_ss.str().c_str());
+  }
+
+  return std::make_pair(graph_with_stats.first,
+                        found_fc_as_y_count + graph_with_stats.second);
+}
+
+GraphWithStats FCResidualConnectionMKLDNNFusePass::FuseProjectionFC(
+    const std::string& name_scope,
+    const GraphWithStats& graph_with_stats) const {
+  GraphPatternDetector gpd;
+  auto pattern = gpd.mutable_pattern();
+
+  patterns::FCMKLDNN fc_x_pattern{pattern, name_scope};
+
+  auto fc_x_output = fc_x_pattern(
+      gpd.mutable_pattern()->NewNode("fc_x")->AsInput()->assert_is_op_input(
+          "fc", "Input"),
+      true);
+
+  patterns::FCMKLDNN fc_y_pattern{pattern, name_scope};
+  auto fc_y_output = fc_y_pattern(
+      gpd.mutable_pattern()->NewNode("fc_y")->AsInput()->assert_is_op_input(
+          "fc", "Input"),
+      true);
+
+  patterns::Elementwise elementwise_pattern{pattern, name_scope};
+  elementwise_pattern(fc_x_output, fc_y_output, "elementwise_add");
+  fc_x_output->AsIntermediate();
+  fc_y_output->AsIntermediate();
+
+  int found_projection_fc_count = 0;
+
+  auto handler = [&](const GraphPatternDetector::subgraph_t& subgraph,
+                     Graph* g) {
+    GET_IR_NODE_FROM_SUBGRAPH(fc_x_op, fc, fc_x_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(fc_x_input, input, fc_x_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(fc_x_weights, weights, fc_x_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(fc_x_output, output, fc_x_pattern);
+
+    GET_IR_NODE_FROM_SUBGRAPH(fc_y_op, fc, fc_y_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(fc_y_input, input, fc_y_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(fc_y_weights, weights, fc_y_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(fc_y_output, output, fc_y_pattern);
+
+    GET_IR_NODE_FROM_SUBGRAPH(elementwise_op, elementwise_op,
+                              elementwise_pattern);
+    GET_IR_NODE_FROM_SUBGRAPH(elementwise_out, elementwise_out,
+                              elementwise_pattern);
+
+    if (HasFusedActivation(fc_x_op) || HasFusedActivation(fc_y_op)) return;
+
+    if (!IsCompat(subgraph, g)) {
+      LOG(WARNING)
+          << "fc_elementwise_add_mkldnn_fuse_pass in op compat failed.";
+      return;
+    }
+
+    if (FindFuseOption(*fc_x_op, *elementwise_op) != FUSE_MKLDNN) return;
+    if (FindFuseOption(*fc_y_op, *elementwise_op) != FUSE_MKLDNN) return;
+
+    Node* projection_node;
+    Node* residual_fc_op;
+    Node* residual_fc_output;
+    if (IsReachable(g, fc_x_input, fc_y_output)) {
+      projection_node = fc_x_output;
+      residual_fc_op = fc_y_op;
+      residual_fc_output = fc_y_output;
+    } else if (IsReachable(g, fc_y_input, fc_x_output)) {
+      projection_node = fc_y_output;
+      residual_fc_op = fc_x_op;
+      residual_fc_output = fc_x_output;
+    } else {
+      return;
+    }
```

Comment:
I can see that many lines of code is duplicated.
Since `A+B == B+A` find a way to reduce it, please.
Maybe make a pass which will enforce that `FC` is always the `X input` for `Add`?